flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/GatherSumApplyITCase.java

```diff
@@ -0,0 +1,212 @@
+		SumFunction<Long, Long, Long, HashSet<Vertex<Long, Long>>> sum =
+				new SumFunction<Long, Long, Long, HashSet<Vertex<Long, Long>>>() {
+
+					@Override
+					public Tuple2<Long, HashSet<Vertex<Long, Long>>> sum(
+							Tuple2<Long, HashSet<Vertex<Long, Long>>> arg0,
+							Tuple2<Long, HashSet<Vertex<Long, Long>>> arg1) {
+
+						HashSet<Vertex<Long, Long>> result = new HashSet<Vertex<Long, Long>>();
+						result.addAll(arg0.f1);
+						result.addAll(arg1.f1);
+
+						return new Tuple2<Long, HashSet<Vertex<Long, Long>>>(arg0.f0, result);
+					}
+				};
+
+		// Find the minimum vertex id in the set which will be propagated
+		ApplyFunction<Long, Long, Long, HashSet<Vertex<Long, Long>>> apply =
+				new ApplyFunction<Long, Long, Long, HashSet<Vertex<Long, Long>>>() {
+
+					@Override
+					public void apply(Tuple2<Long, HashSet<Vertex<Long, Long>>> set,
+							Vertex<Long, Long> src, Collector<Vertex<Long, Long>> collector) {
+						long minValue = src.getValue();
+						for (Vertex<Long, Long> v : set.f1) {
+							if (v.getValue() < minValue) {
+								minValue = v.getValue();
+							}
+						}
+
+						// This is the condition that enables the termination of the iteration
+						if (minValue != src.getValue()) {
+							collector.collect(new Vertex<Long, Long>(src.getId(), minValue));
+						}
+					}
+				};
+
+		Graph<Long, Long, Long> minColoring = graph.runGatherSumApplyIteration(gather, sum, apply, 16);
+		minColoring.getVertices().writeAsCsv(resultPath);
+
+		env.execute();
+
+		expectedResult = "1,1\n" +
+				"2,1\n" +
+				"3,1\n" +
+				"4,1\n" +
+				"5,1\n";
+	}
+
+	@Test
+	public void testSingleSourceShortestPath() throws Exception {
+		/*
+		 * This test runs the Single Source Shortest Path algorithm using a GSA iteration
+		 */
+		final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		// Start off with 0 for the source vertex and INF for all other vertices
+		List<Vertex<Long, Double>> vertices = new ArrayList<Vertex<Long, Double>>();
+		vertices.add(new Vertex<Long, Double>(1L, 0.0));
+		vertices.add(new Vertex<Long, Double>(2L, Double.POSITIVE_INFINITY));
+		vertices.add(new Vertex<Long, Double>(3L, Double.POSITIVE_INFINITY));
+		vertices.add(new Vertex<Long, Double>(4L, Double.POSITIVE_INFINITY));
+		vertices.add(new Vertex<Long, Double>(5L, Double.POSITIVE_INFINITY));
+
+		Graph<Long, Double, Long> graph = Graph.fromDataSet(env.fromCollection(vertices),
+				TestGraphUtils.getLongLongEdgeData(env), env);
+
+		// The path from src to trg through edge e costs src + e
+		// If the target's distance is 0 (the target is the actual source), return 0 instead
+		GatherFunction<Long, Double, Long, Double> gather =
+				new GatherFunction<Long, Double, Long, Double>() {
+					@Override
+					public Tuple2<Long, Double> gather(Tuple3<Vertex<Long, Double>,
+							Edge<Long, Long>, Vertex<Long, Double>> triplet) {
+
+						if (triplet.f2.getValue() == 0.0) {
```

Comment:
Normally this is not okay, but we never change the initial 0.0 value.
Should we still make an epsilon check here, even though it would result in (a bit) less readable code?